Ensure that Rule.formula is serializable

IagoAbal opened the pull request:
The `Regexp_engine.t` type is a wrapper for a C type coming from the C PCRE library, and it is represented by a "Custom" block. Values of this type are not serializable, and will cause an exception if we try calling `Marshal.to_string` (or variant) on them. This can cause problems when serializing any data structure that depends on `Xpattern.t` and `Rule.formula`.

Previously, we got weird crashes after PR #5725 (taint labels). This was due to redefining type `Taint.source` as `Rule.taint_source call_trace`, where `Rule.taint_source` contains a `Rule.formula`. Whenever a taint finding is produced, we compute the taint trace this way:

    lazy (taint_trace_of_src_to_sink source tokens sink)

This `lazy ...` expression creates a thunk that capturies a `source` which contains a `Rule.formula`. Even though the formula is not used to compute the taint trace, it still needs to be serialized in order to serialize the thunk.

So, when using -j 2 (or higher) and -json (in text mode we print findings right away and hence the thunk gets evaluated), this thunk remains unevaluated and reaches the result of `Match_rules.check`. Then Parmap tries to serialize it to communicate the result back to the master process... and if the formula contains any compiled regex, it crashes.

This also affects us if we try to serialize a `Dataflow_tainting.config` for caching purposes. While we could remove the formulas from these types, it is less error prone to just make sure that formulas can be serialized.

Follows: 2e6b3079d51 ("Quick-fix for regression in 0.107.0 after taint labels PR (#5867)")

Regarding perf, this should not cause any noticeable regression, unless we build some synthetic case with lots of rules making an absurd use of regexes. I ran `semgrep -c p/default` on three repos from stress-test-monorepo (elasticsearch, vscode, and mediawiki) and with this change all runs were actually a little bit faster (maybe just coincidence).

test plan:
make test
(In PR #5867 we modified Test_engine.ml to check that the result of Match_rules.check can be serialized, see 'Test_engine.make_tests'.)

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

aryx commented on semgrep-core/src/engine/Match_search_mode.ml:
```diff
@@ -124,7 +124,8 @@ let partition_xpatterns xs =
          | XP.Spacegrep x -> Common.push (x, pid, str) spacegrep
-         | XP.Regexp x -> Common.push (x, pid, str) regexp
+         | XP.Regexp x ->
+             Common.push (Regexp_engine.pcre_compile x, pid, str) regexp
```
you can add a [@@profiling] on Regexp_engine.pcre_compile so we can track (thx to -profile) whether
those calls are expensive.

IagoAbal replied:
Good point!

aryx commented on semgrep-core/src/core/Xpattern.ml:
```diff
@@ -21,15 +21,30 @@
 
-type regexp = Regexp_engine.t [@@deriving show, eq]
+type compiled_regexp = Regexp_engine.t [@@deriving show, eq]
+type regexp_string = string [@@deriving show, eq]
```
maybe a comment here like '(* see the NOTE below on Regexp for the need to have this type *)'

aryx reviewed (approved):
Would have been great to have a test case showing the issue.
Otherwise looks good.

IagoAbal commented:
> Would have been great to have a test case showing the issue.
> Otherwise looks good.

Oh we do have that, see `Test_engine.ml`:
```ocaml
             (* Check that the result can be marshalled, as this will be needed
                 when using Parmap! See PA-1724. *)
             (try Marshal.to_string res [ Marshal.Closures ] |> ignore with
             | exn ->
                 failwith
                   (spf "exn on %s (exn = %s)" file (Common.exn_to_s exn)));
```
If you undo this change (and keep the `lazy` in Match_tainting_mode` you get many failing tests).

IagoAbal committed:
Ensure that Rule.formula is serializable

The `Regexp_engine.t` type is a wrapper for a C type coming from the
C PCRE library, and it is represented by a "Custom" block. Values of
this type are not serializable, and will cause an exception if we try
calling `Marshal.to_string` (or variant) on them. This can cause
problems when serializing any data structure that depends on `Xpattern.t`
and `Rule.formula`.

Previously, we got weird crashes after PR #5725 (taint labels). This was
due to redefining type `Taint.source` as `Rule.taint_source call_trace`,
where `Rule.taint_source` contains a `Rule.formula`. Whenever a taint
finding is produced, we compute the taint trace this way:

    lazy (taint_trace_of_src_to_sink source tokens sink)

This `lazy ...` expression creates a thunk that capturies a `source`
which contains a `Rule.formula`. Even though the formula is not used to
compute the taint trace, it still needs to be serialized in order to
serialize the thunk.

So, when using -j 2 (or higher) and -json (in text mode we print findings
right away and hence the thunk gets evaluated), this thunk remains
unevaluated and reaches the result of `Match_rules.check`. Then Parmap
tries to serialize it to communicate the result back to the master
process... and if the formula contains any compiled regex, it crashes.

This also affects us if we try to serialize a `Dataflow_tainting.config`
for caching purposes. While we could remove the formulas from these
types, it is less error prone to just make sure that formulas can be
serialized.

Follows: 2e6b3079d51 ("Quick-fix for regression in 0.107.0 after taint labels PR (#5867)")

Regarding perf, this should not cause any noticeable regression, unless
we build some synthetic case with lots of rules making an absurd use of
regexes. I ran `semgrep -c p/default` on three repos from stress-test-monorepo
(elasticsearch, vscode, and mediawiki) and with this change all runs were
actually a little bit faster (maybe just coincidence).

test plan:
make test
(In PR #5867 we modified Test_engine.ml to check that the result of
 Match_rules.check can be serialized, see 'Test_engine.make_tests'.)

IagoAbal committed:
Yoann's review

IagoAbal commented:
I wonder what is going on, tests passed 5 days ago, semgrep-core tests pass locally. I rebased against `develop` and trying again...

aryx commented:
I did modify some code recently in Parse_rule.ml regarding regexps in parse_xptattern.

aryx commented:
That's the CI failure. See the error in Parse_rule.ml

IagoAbal committed:
Rebase against develop

IagoAbal commented:
Yeah, I was running the tests before rebasing so I didn't get that error. I got it when I rebased.
